Grbl_Esp32/tools/SpindleClass.cpp

```diff
@@ -0,0 +1,403 @@
+/*
+    SpindleClass.cpp
+
+    A Spindle Class
+        Spindle         - A base class. Do not use
+        PWMSpindel      - A spindle with a PWM output
+        RelaySpindle    - An on/off only spindle
+        Laser           - Output is PWM, but the M4 laser power mode can be used
+        DacSpindle      - Uses the DAC to output a 0-3.3V output
+
+    Part of Grbl_ESP32
+
+    2020 -	Bart Dring
+
+    Grbl is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    Grbl is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+    You should have received a copy of the GNU General Public License
+    along with Grbl.  If not, see <http://www.gnu.org/licenses/>.
+
+    TODO
+        Consider breaking into one file per class.
+
+    Get rid of dependance on machine definition #defines
+        SPINDLE_PWM_BIT_PRECISION
+        SPINDLE_PWM_PIN
+        SPINDLE_ENABLE_PIN
+        SPINDLE_DIR_PIN
+
+*/
+#include "grbl.h"
+#include "SpindleClass.h"
+
+bool Spindle::isRateAdjusted() {
+    return false; // default for basic spindles is false
+}
+
+void Spindle :: spindle_sync(uint8_t state, float rpm) {
```

Comment:
the .spindle_sync() method does not appear to be used anywhere.  I suspect that you meant to delete this after creating .set_state().

Reply:
It is used in the base class because all spindles need it. It basically calls `protocol_buffer_synchronize();` to make sure spindle stuff is synchronized with gcode.